File: Reverse Nodes in k-Group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * @param head a ListNode
     * @param k an integer
     * @return a ListNode
     */
    ListNode *reverseKGroup(ListNode *head, int k) {
        // Write your code here
        ListNode dummy(INT_MIN);
        dummy.next = head;
        ListNode *h = &dummy;
        while (h)
        {
            ListNode *first = h;
            ListNode *last = first;
            int loops = k;
            while (last && loops > 0)
            {
                --loops;
                last = last->next;
            }
            if (!last)
            {
                return dummy.next;
            }
            else
            {
                ListNode *pre = first->next;
                h = pre;
                int loops = k;
                while (loops > 1)
                {
                    --loops;
                    ListNode *next = pre->next;
                    pre->next = last->next;
                    last->next = pre;
                    first->next = next;
                    pre = next;
                }
            }
            
        }
        return dummy.next;
        
    }
};

File: Reverse Nodes in k-Group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @param k, an integer
    # @return a ListNode
    def reverseKGroup(self, head, k):
        # Write your code here
        if not head or k == 0:
            return head
        
        cur = head
        length = 0
        while cur:
            length += 1
            cur = cur.next
        times = length / k
        dummy = ListNode(0)
        # cur, end = head, dummy
        # for __ in range(times):
        #     next, tail, cur = cur, cur, cur.next
        #     next.next = None
        #     for __ in range(k - 1):
        #         cur.next, cur, next = next, cur.next, cur
        #     end.next, end = next, tail
        # end.next = cur
        tail = dummy
        cur = head
        for __ in range(times):
            next = cur
            for __ in range(k):
                tail.next, cur.next, cur = cur, tail.next, cur.next
            tail = next
        tail.next = cur
        return dummy.next
            
            